packages/@aws-cdk/aws-iam/lib/principals.ts

```diff
@@ -154,8 +154,11 @@ export class PrincipalWithConditions implements IPrincipal {
   public readonly assumeRoleAction: string = this.principal.assumeRoleAction;
   private additionalConditions: Conditions;
 
+  /**
+   * @param principal The underlying principal to which conditions will be attached.
+   */
   constructor(
-    private readonly principal: IPrincipal,
+    public readonly principal: IPrincipal,
```

Comment:
In order to parse the correct principal, we need to unwrap the underlying principal to access the condition-free service or AWS account this principal represents. That is, if we don't unwrap `new iam.PrincipalWithConditions(new iam.ServicePrincipal('my-service'), ...)`, the `if ('service' in principal)` conditional fails instead of returning `'my-service'`.

Instead of exposing the underlying principal, we could do the following in `parsePermissionPrincipal`:
```
// eslint-disable-next-line dot-notation
principal = principal['principal'];
```
This would be circumventing the type system, which we're already doing with duck-typing in this function.

Reply:
Not the best solution but I suppose that'll do.
Something around how the various `Principal` types are modeled feels off here, but that's probably larger change than this requires.